syntax = "proto3";
option cc_enable_arenas = true;

package NKqpProto;
option java_package = "ru.yandex.kikimr.proto";

import "ydb/library/yql/dq/actors/protos/dq_stats.proto";

message TKqpStatsCompile {
    bool FromCache = 1;
    uint64 DurationUs = 2;
    uint64 CpuTimeUs = 3;
}

// Per datashard table extra stats
message TKqpShardTableExtraStats {
    uint64 ShardId = 1;
    uint64 ShardCpuTimeUs = 2;
}

// aggregated shard extra stats for table
message TKqpShardTableAggrExtraStats {
    uint32 AffectedShards = 1;
    NYql.NDqProto.TDqStatsAggr ShardCpuTimeUs = 2;
}

// aggregated read actor extra stats for table
message TKqpReadActorTableAggrExtraStats {
    repeated uint64 AffectedShards = 1;
}

// extra for NYql.NDqProto.TDqTableStats
message TKqpTableExtraStats {
    TKqpReadActorTableAggrExtraStats ReadActorTableAggrExtraStats = 1;
}

// extra for NYql.NDqProto.TDqComputeActorStats
message TKqpComputeActorExtraStats {
}

// extra for NYql.NDqProto.TDqStageStats
message TKqpStageExtraStats {
    repeated NYql.NDqProto.TDqTaskStats DatashardTasks = 1;
    repeated TNodeExecutionStats NodeStats = 4;
}

message TKqpScanTaskExtraStats {
    uint32 RetriesCount = 1;
}

// extra for NYql.NDqProto.TDqTaskStats
message TKqpTaskExtraStats {
    TKqpScanTaskExtraStats ScanTaskExtraStats = 1;
}

message TNodeExecutionStats {
    optional uint32 NodeId = 1;
    optional uint32 ShardsCount = 2;
}

// extra for NYql.NDqProto.TDqExecutionStats
message TKqpExecutionExtraStats {
    // Basic stats
    uint32 AffectedShards = 1;
    NYql.NDqProto.TDqStatsAggr ComputeCpuTimeUs = 2;   // compute actors time
    NYql.NDqProto.TDqStatsAggr ShardsCpuTimeUs = 3;    // datashards time

    // Profile stats
    uint64 ResolveCpuTimeUs = 100;
    uint64 ResolveWallTimeUs = 101;
}

message TKqpStatsQuery {
    // Basic stats
    uint64 DurationUs = 1;
    uint64 QueuedTimeUs = 9;
    TKqpStatsCompile Compilation = 2;

    reserved 3; // repeated TKqpStatsExecution Executions = 3;

    uint64 WorkerCpuTimeUs = 4;
    uint64 ReadSetsCount = 5;
    uint64 MaxShardProgramSize = 6;
    uint64 MaxShardReplySize = 7;

    repeated NYql.NDqProto.TDqExecutionStats Executions = 8;
}
